### DISCUSSION TITLE
Welcome to LogixLibraries Discussions!

### DISCUSSION THREAD

**Discussion by JeremyMedders** [JeremyMedders/LogixLibraries#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MickeyM79** [comment]
Stumbled across this browsing the PLCTalk forum.  Very cool idea.  Do you have any guidelines for contributing?  I'm happy to contribute as I have time. Granted, these days I don't have an overabundance of time.

**Comment by JeremyMedders** [comment]
> Stumbled across this browsing the PLCTalk forum.  Very cool idea.  Do you have any guidelines for contributing?  I'm happy to contribute as I have time. Granted, these days I don't have an overabundance of time.

Thank you Mickey and welcome! No pressure on contributions or particular timelines. That kind of thing is pretty much self-imposed on me and the place is largely an open book this early on.

I'm making continuous changes to the repo structure and am pushing code up and I crunch it out and test.

Some basic guidelines I'm using include tracking which firmware version and hardware are used to build each instruction and/or sample code. I encourage code in general to be as portable as possible but at times it's simply out of reach due to artificially imposed limitations between Logix families. That's okay, and it just needs to be noted.

One thing I'm counting on everyone to stick to is exporting to the .L5X format.

Documentation in the form of PDF, MS word docs, .txt files, Excel spreadsheets, and so on are all acceptable.

**Comment by TheColonel2688** [comment]
Looking forward to more documentation on stuff dealing with CIP and Messaging.

Dumb question could you define what the Cmd prefixes mean

What is PCmd, Mcmd, Xcmd, etc. 

I tried to find Pax documentation talking about thst but I have found nothing on Google.

**Reply by JeremyMedders** [reply]
What kinds of CIP/messaging documentation are you interested in seeing? It's its own universe, so I'm hoping to lasso as much of what people actually want to see within these utilities.

**Reply by TheColonel2688** [reply]
@JeremyMedders 

I have a "Library" I maintain that we give to our customers. This library basically wraps all of the remote interaction with our machines in a nice "happy meal" for them. They put this code on their main plant PLCs so they can remotely control our machines over EIP.

So I use the MSG instruction in there. I can have multiple "Instances" of this "Library" (Basically a "Program" that is a template with my own AOIs in it), and each instance needs to target different tags on our PLC. So my biggest headache is making sure the Remote Paths (Tags paths on our PLC) in the MSG UDTs are correct for each instance.

(I can send you a copy of my "Library" if you it would help you understand.)

I would like to make it so I can just specify in the code what the Remote **Tag** Path String is and what the Remote IP **address** is.

I use CIP Data Table Read and CIP Data Table Write. I have tried writing the PLC address in the past, and I had it working on a 5**3**70. but when I moved it to a Customers 5**5**70 it didn't work anymore. I have no idea why. I haven't had more time to fiddle with it.

So if I could do something like, Path `API_Thing{ThingInstance}_WriteSlot[{WriteSlot}]`
So then for 
```
ThingInstance ="A"
WriteSlot = "5"
```

It would make the remote tag path
`API_ThingA_WriteSlot[5]`

But really, I can handle the string concatenation; I just want an AOI that will write the strings correctly to the MSG Instruction Backing UDT.

Sidenote: a string "Interpolator" would be awesome

Regarding more generalized CIP documentation, I have never found anything that gave a nice walkthrough of how CIP (EIP) works etc.

I also find things confusing between what is supported in the official standard and what is Allen Bradley only.

It seems that symbolic address is part of the CIP standard, but that is not the same as Allen Bradley Tag names.

Does that help?

**Reply by JeremyMedders** [reply]
I can look at it if that's okay. I've learned some pathing tricks along the way that I might be able to show you.

**Reply by TheColonel2688** [reply]
@JeremyMedders I sent you a PM with my email address on PLCTalk.com since GitHub doesn't have private messaging.

**Reply by JeremyMedders** [reply]
I replied to you there. My email is good.

**Comment by JeremyMedders** [comment]
Yes sir, wiki here under construction as of yesterday.

- PCmd: commands from program
- OCmd: commands from operator
- MCmd: commands from maintenance
- XCmd: commands from external source (scada)

**Reply by TheColonel2688** [reply]
What is the difference between OCmd and XCmd? The operator would only be Using an HMI right? In this context, is SCADA any different than an HMI? 

I'm just a curious person.

**Reply by JeremyMedders** [reply]
OCmd could originate from a local HMI and an XCmd from a remote one. Another aspect is that permissions can be assigned such that an operator be might able to jog a motor in reverse to clear a jam but an external source can't. An external source could also be configured to not "exist" (or vice-versa) in the example you provided.

**Comment by TheColonel2688** [comment]
Hmmm that makes sense

**Comment by TheColonel2688** [comment]
@JeremyMedders, could you add the license to the readme? Legally people need a license to use this code. 

I would prefer MIT, but MPL 2.0 is ok, so it is compatible with Commerical Code.

(GPLs "Commerical" Compatibility is a big lie, Richard Stallman is a Political activist and nothing more, GPL licenses are Cancerous) 

here is an interesting Discussion thread, on the RustLang Forums

https://users.rust-lang.org/t/licensing-mit-apache-2-vs-mpl-2-0/46250

**Reply by TheColonel2688** [reply]
Also, I know the 5x70 series is Legacy now, but it would be an nice addition to the library to add a String Copy AOI that is compatible with the 5x70 series.

Trying to deal with string in 5x70 is terrible.

**Reply by JeremyMedders** [reply]
> @JeremyMedders, could you add the license to the readme? Legally people need a license to use this code.
> 
> I would prefer MIT, but MPL 2.0 is ok, so it is compatible with Commerical Code.
> 
> (GPLs "Commerical" Compatibility is a big lie, Richard Stallman is a Political activist and nothing more, GPL licenses are Cancerous)
> 
> here is an interesting Discussion thread, on the RustLang Forums
> 
> https://users.rust-lang.org/t/licensing-mit-apache-2-vs-mpl-2-0/46250

I'll start with an MIT license for now. It's been a to-do of mine but since others are now asking, it's probably time to attach it.

I will also do so in the opening comments of code blocks.

**Reply by JeremyMedders** [reply]
> Also, I know the 5x70 series is Legacy now, but it would be an nice addition to the library to add a String Copy AOI that is compatible with the 5x70 series.
> 
> Trying to deal with string in 5x70 is terrible.

By "string copy" are you referring to the ability in a 5x80 project to do either of these:
- `MOV Str1 Str2` in ladder or
- `Str1 := Str2; ` or `Str1 := 'literal';` in structured text?

I also happen to 100% agree with you - 5x70 limitations in modern versions of Studio (v28+) are staggering and Rockwell commits resources to handling the workaround logic in their own AOIs over what amounts to an artificial limitation to sell more modern hardware.

Don't get me started on how they impose garbage limitations on unsigned numbers. The idea that a fundamental concept in computer science rests on a sales divide in 2023 is as obnoxious as it gets. The Common Industrial Protocol has used unsigned values since its inception and having to ham-fist workarounds when parsing CIP data in a 5x70 is doubly annoying. Who at Rockwell hates their users? 

My opinion is that the ability to do a String literal assignment or deal with unsigned numbers, or basically construct a piece of code, should rest with the version of software you have, not the hardware. We've all ponied up year over year for the new versions and should be able to deploy the features they include everywhere, unless there's an actual hardware constraint (which doesn't exist).

**Reply by JeremyMedders** [reply]
All that said,

A general Str_Copy AOI would be possible and work on both hardware families. The underlying data in a Logix String is DINT and SINT, so the use of assignment or COP upon these nested members anywhere in code shouldn't be an issue.

**Reply by TheColonel2688** [reply]
> By "string copy" are you referring to the ability in a 5x80 project to do either of these:
> 
>     * `MOV Str1 Str2` in ladder or
> 
>     * `Str1 := Str2; ` or `Str1 := 'literal';` in structured text?
> 

'MOV Str1, Str2' in ladder. I don't think you can do a 'Str1 := Str2' either can you?

mostly 

I know string literals are only 5380. 

I have had issues with using COP by itself to copy strings because I don't think it updates the LEN member automatically.

Another issue is no Length Checking.

Yes, you can use bare COP instruction, but it is more opportunity to introduce bugs.

anyway, these are just my thoughts. 

I have been heavily into Codesys the last year or some, so I haven't been doing much in Rockwell. Love codesys makes me never want to touch Rockwell again lol


> I also happen to 100% agree with you - 5x70 limitations in modern versions of Studio (v28+) are staggering and Rockwell commits resources to handling the workaround logic in their own AOIs over what amounts to an artificial limitation to sell more modern hardware.
> 
> Don't get me started on how they impose garbage limitations on unsigned numbers. The idea that a fundamental concept in computer science rests on a sales divide in 2023 is as obnoxious as it gets. The Common Industrial Protocol has used unsigned values since its inception and having to ham-fist workarounds when parsing CIP data in a 5x70 is doubly annoying. Who at Rockwell hates their users?
> 
> My opinion is that the ability to do a String literal assignment or deal with unsigned numbers, or basically construct a piece of code, should rest with the version of software you have, not the hardware. We've all ponied up year over year for the new versions and should be able to deploy the features they include everywhere, unless there's an actual hardware constraint (which doesn't exist).

I completely agree. I have had some Rockwell Fan Boys try to "Explain" to me how their older hardware can't support unsigned types and string literals before.....

They don't seem to know that that has nothing to do with hardware. I assume Instructions in an ACD are not compiled to bare metal CPU instructions. they are compiled into a runtime-level language, and the firmware knows how to execute that runtime-level language.

Even if that is not true, Unsigned types and string literals would still be a compiler issue. 

I think Rockwell probably went out of their way to impose this restriction on 5x70 systems as a way to try to get plants to upgrade and for all new machines to use their new 80 series.

Talking with one of their Firmware developers once, he said that Engineering is not allowed to make any improvements that Sales doesn't approve, and Sales only approves things if they project that it will make them more money.

**Reply by JeremyMedders** [reply]
That's unfortunate, in a big way, I think.

That said, you should have a Str_Copy function to try. Should be able to handle all the possible forms of a Logix String.

**Reply by TheColonel2688** [reply]
> That's unfortunate, in a big way, I think.
> 
> That said, you should have a Str_Copy function to try. Should be able to handle all the possible forms of a Logix String.

Oh thanks, I hadn't noticed the new commit.

I will check it out 

Thanks!

**Comment by TheColonel2688** [comment]
What is the parameter lib you added? I haven't had a chance to open it in Studio and look

**Reply by JeremyMedders** [reply]
It's an early attempt at making the basic types changeable online by allowing them to be cast to other types - the core PAR_ANY type is aligned to 8 bytes so it will work with whatever gets stuffed in it.

Par_Cast has the added perk of "CanBeNull" InOut parameters, so you can cast any source type to PAR, or PAR to another PAR, without needing dummy values eating memory for AOI-required parameters.

Pretty inefficient for BOOL, but the ability to store heterogenous data in the same array without resorting to gigantic monolith UDTs might be useful.

PAR_ANY is a 64-bit overlay, so one could cast REAL/LREAL into it and be able to manipulate the bits of floating-point data (and of course the other numeric types) in a uniform way.

The lofty goal is a recipe manager that permits its users to not only set setpoints/limits but _the data type as well_.

Time will tell...

**Comment by jhtressl** [comment]
Hi. Cool project! What do you have in mind for profiling?

**Reply by JeremyMedders** [reply]
Hi and thanks!

What I'd like to eventually have is a database of the library instructions / sample programs, how much memory they consume, the max duration of a scan cycle, etc. for different processors (5380/5480/5580).

**Comment by Moosetracks555** [comment]
Hello, 
I was looking for a PF525 AOI and someone on reddit pointed me here.  I got the AOI imported fine.  I thought the l5x files might be example code, but I can't get those to import or open.  Is this an AOI you built, or is it something from PlantPax?  Is there any documentation on what all the tags are?

**Reply by JeremyMedders** [reply]
Hi Moose,

An L5X can be a program, an AOI, a UDT, an IO connection, or even an entire project.

What I recommend is that you first import the IO module in the 525 folder (you can change the IP later) and then import the entire sample program. It will include other "Op" AOIs. The PF525's parameters are documented on the block and the naming is modeled after PAx, but I wrote these myself.

Hope this helps and keep the questions coming if you need help :)

**Comment by JeremyMedders** [comment]
> Hello, 
> 
> I was looking for a PF525 AOI and someone on reddit pointed me here.  I got the AOI imported fine.  I thought the l5x files might be example code, but I can't get those to import or open.  Is this an AOI you built, or is it something from PlantPax?  Is there any documentation on what all the tags are?

Hi Moose,

An L5X can be a program, an AOI, a UDT, an IO connection, or even an entire project.

What I recommend is that you first import the IO module in the 525 folder (you can change the IP later) and then import the entire sample program. It will include other "Op" AOIs. The PF525's parameters are documented on the block and the naming is modeled after PAx, but I wrote these myself.

Hope this helps and keep the questions coming if you need help :)

**Comment by ApolloPLC** [comment]
New to GH.
How would you like people to report a bug and solution?

**Comment by JeremyMedders** [comment]
Hi ApolloPLC,You can do it here or on the discussions pages by starting a thread there.I appreciate your feedback!Sent from my iPhoneOn Aug 17, 2023, at 10:43, ApolloPLC ***@***.***> wrote:﻿
New to GH.
How would you like people to report a bug and solution?

**Reply by JeremyMedders** [reply]
Hi ApolloPLC,

You can do it here or on the discussions pages by starting a new thread.

I appreciate your feedback!
